### DISCUSSION TITLE
Multi-channels image random crop

### DISCUSSION THREAD

**Discussion by p71-23** [Project-MONAI/MONAI#7621] [General]
Does monai have a cropping method for ct\pet multi-channel input? I've only found the RandSpatialCropSamplesd function so far, and it seems that it can only handle single-channel images and the cropping of the corresponding labels.

Error when inputting multiple channels: image and label dimensions do not match.

**Comment by KumoLiu** [comment]
Hi @p71-23, thanks for raising this one.

Could you please create a small piece of code that can help reproduce the issue? Not sure which check cause the issue.
The alternative that jumped into my head at the moment is to do it with `SplitDimd` and `ConcatItemsd`.

Thanks!

**Comment by p71-23** [comment]
Thanks for your warmly response!
I stack CT and PET images into a combined image, like this:

class CustomDataset(Dataset):
def init(self, ct_list, pet_list, mask_list):
self.ct_list = ct_list
self.pet_list = pet_list
self.mask_list = mask_list

def __len__(self):
    return len(self.ct_list)

def __getitem__(self, idx):
    ct_image = nib.load(self.ct_list[idx]).get_fdata()
    ct_image = np.expand_dims(ct_image, axis=0)
    ct_image = ct_image.astype(np.float32)
    pet_image = nib.load(self.pet_list[idx]).get_fdata()
    pet_image = np.expand_dims(pet_image, axis=0)
    pet_image = pet_image.astype(np.float32)
    combined_image = np.stack([ct_image, pet_image], axis=-1)
    mask_image = nib.load(self.mask_list[idx]).get_fdata()
    mask_image = np.expand_dims(mask_image, axis=0)
    mask_image = (mask_image != 0).astype(np.float32)
    return {"image": combined_image, "label": mask_image}

Then, I utilize it as follows:

train_transforms = Compose([
RandCropByPosNegLabeld(
keys=["image", "label"],
label_key="label",
spatial_size=(96, 96, 96),
pos=1,
neg=1,
num_samples=3,
image_key="image",
image_threshold=0,
),
])
train_ds = Dataset(data=train_dataset, transform=train_transforms)
train_loader = DataLoader(train_ds, batch_size=1)

However, I encountered the following error:

Traceback (most recent call last):
File "...", line ..., in apply_transform
...
ValueError: operands could not be broadcast together with shapes (571473920,) (285736960,)
The size (571473920,) is the sum of CT and PET images, while the label is just one channel (285736960,).

I was pleasantly surprised to find that the RandSpatialCropSamplesd function is able to solve this problem, although it can't split based on proportionality like RandCropByPosNegLabeld, but it seems to be able to deal with multi-channel cropping (I'm not really sure, since the final result hasn't been calculated yet)

**Comment by KumoLiu** [comment]
Hi @p71-23,  I believe `RandCropByPosNegLabeld` can work with multi-channel, try the code below.
```
import torch
from monai.transforms import RandCropByPosNegLabeld


img = torch.rand([3, 96, 96])
lbl = torch.randint(0, 3, [1, 96, 96])


trans = RandCropByPosNegLabeld(["img", 'lbl'], "lbl", spatial_size=(32, 32))
out = trans({"img": img,  "lbl": lbl})
```

Thanks.
